File: CS/CustomProgressIndicator/MyProgressIndicationService.cs
using System;
using System.Threading;
using System.Windows.Forms;
using DevExpress.Snap;
using DevExpress.Snap.Core.Services;
using DevExpress.Utils;
using DevExpress.XtraEditors;
// ...

namespace CustomProgressIndicator {
    class MyProgressIndicationService : ISnapMailMergeProgressIndicationService, IDisposable {
        readonly SnapControl snapControl;
        MyProgressIndicationForm frm;
        CancellationTokenSource cts;
        CancellationToken token;

        public MyProgressIndicationService(SnapControl snapControl) {
            Guard.ArgumentNotNull(snapControl, "snapControl");
            this.snapControl = snapControl;
        }

        #region ISnapMailMergeProgressIndicationService Members

        public CancellationToken CancellationToken { get { return token; } }

        public void Reset() {
            if (!object.ReferenceEquals(this.cts, null)) {
                this.cts.Dispose();
                this.cts = null;
            }
            this.cts = new CancellationTokenSource();
            this.token = this.cts.Token;
        }

        #endregion

        #region IProgressIndicationService Members

        public void Begin(string displayName, int minProgress, int maxProgress,
            int currentProgress) {
            Action begin = delegate {
                this.frm = new MyProgressIndicationForm();
                SubscribeFormEvents();
                this.frm.Begin(minProgress, maxProgress, currentProgress);
                this.frm.ShowDialog(this.snapControl);
            };
            this.snapControl.BeginInvoke(begin);
        }

        public void End() {
            if (object.ReferenceEquals(this.frm, null))
                return;
            Action end = delegate {
                UnsubscribeFormEvents();
                this.frm.Dispose();
                this.frm = null;
            };
            this.snapControl.BeginInvoke(end);
        }

        public void SetProgress(int currentProgress) {
            if (object.ReferenceEquals(this.frm, null))
                return;
            Action setProgress = delegate {
                this.frm.SetProgress(currentProgress);
            };
            this.snapControl.BeginInvoke(setProgress);
        }

        #endregion

        void SubscribeFormEvents() {
            this.frm.FormClosing += OnProgressIndicationFormClosing;
        }

        void UnsubscribeFormEvents() {
            if (object.ReferenceEquals(this.frm, null))
                return;
            this.frm.FormClosing -= OnProgressIndicationFormClosing;
        }

        void OnProgressIndicationFormClosing(object sender, FormClosingEventArgs e) {
            if (XtraMessageBox.Show(this.frm.LookAndFeel, this.frm,
                "Do you really want to interrupt mail merge?", Application.ProductName,
                MessageBoxButtons.YesNo) == DialogResult.Yes)
                this.cts.Cancel();
            else
                e.Cancel = true;
        }

        #region IDisposable Members

        public void Dispose() {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        ~MyProgressIndicationService() {
            Dispose(false);
        }
        protected virtual void Dispose(bool disposing) {
            if (disposing) {
                if (!object.ReferenceEquals(this.frm, null)) {
                    this.frm.Dispose();
                    this.frm = null;
                }
                if (!object.ReferenceEquals(this.cts, null)) {
                    this.cts.Dispose();
                    this.cts = null;
                }
            }
        }

        #endregion
    }
}

File: VB/CustomProgressIndicator/MyProgressIndicationService.vb
Imports System
Imports System.Threading
Imports System.Windows.Forms
Imports DevExpress.Snap
Imports DevExpress.Snap.Core.Services
Imports DevExpress.Utils
Imports DevExpress.XtraEditors

' ...
Namespace CustomProgressIndicator

    Friend Class MyProgressIndicationService
        Implements ISnapMailMergeProgressIndicationService, IDisposable

        Private ReadOnly snapControl As SnapControl

        Private frm As MyProgressIndicationForm

        Private cts As CancellationTokenSource

        Private token As CancellationToken

        Public Sub New(ByVal snapControl As SnapControl)
            Guard.ArgumentNotNull(snapControl, "snapControl")
            Me.snapControl = snapControl
        End Sub

'#Region "ISnapMailMergeProgressIndicationService Members"
        Public ReadOnly Property CancellationToken As CancellationToken Implements ISnapMailMergeProgressIndicationService.CancellationToken
            Get
                Return token
            End Get
        End Property

        Public Sub Reset() Implements ISnapMailMergeProgressIndicationService.Reset
            If Not ReferenceEquals(cts, Nothing) Then
                cts.Dispose()
                cts = Nothing
            End If

            cts = New CancellationTokenSource()
            token = cts.Token
        End Sub

'#End Region
'#Region "IProgressIndicationService Members"
        Public Sub Begin(ByVal displayName As String, ByVal minProgress As Integer, ByVal maxProgress As Integer, ByVal currentProgress As Integer) Implements DevExpress.Services.IProgressIndicationService.Begin
            Dim lBegin As Action = Sub()
                frm = New MyProgressIndicationForm()
                SubscribeFormEvents()
                frm.Begin(minProgress, maxProgress, currentProgress)
                frm.ShowDialog(snapControl)
            End Sub
            snapControl.BeginInvoke(lBegin)
        End Sub

        Public Sub [End]() Implements DevExpress.Services.IProgressIndicationService.End
            If ReferenceEquals(frm, Nothing) Then Return
            Dim lEnd As Action = Sub()
                UnsubscribeFormEvents()
                frm.Dispose()
                frm = Nothing
            End Sub
            snapControl.BeginInvoke(lEnd)
        End Sub

        Public Sub SetProgress(ByVal currentProgress As Integer) Implements DevExpress.Services.IProgressIndicationService.SetProgress
            If ReferenceEquals(frm, Nothing) Then Return
            Dim lSetProgress As Action = Sub() frm.SetProgress(currentProgress)
            snapControl.BeginInvoke(lSetProgress)
        End Sub

'#End Region
        Private Sub SubscribeFormEvents()
            AddHandler frm.FormClosing, AddressOf OnProgressIndicationFormClosing
        End Sub

        Private Sub UnsubscribeFormEvents()
            If ReferenceEquals(frm, Nothing) Then Return
            RemoveHandler frm.FormClosing, AddressOf OnProgressIndicationFormClosing
        End Sub

        Private Sub OnProgressIndicationFormClosing(ByVal sender As Object, ByVal e As FormClosingEventArgs)
            If XtraMessageBox.Show(frm.LookAndFeel, frm, "Do you really want to interrupt mail merge?", Application.ProductName, MessageBoxButtons.YesNo) = DialogResult.Yes Then
                cts.Cancel()
            Else
                e.Cancel = True
            End If
        End Sub

'#Region "IDisposable Members"
        Public Sub Dispose() Implements IDisposable.Dispose
            Dispose(True)
            GC.SuppressFinalize(Me)
        End Sub

        Protected Overrides Sub Finalize()
            Dispose(False)
        End Sub

        Protected Overridable Sub Dispose(ByVal disposing As Boolean)
            If disposing Then
                If Not ReferenceEquals(frm, Nothing) Then
                    frm.Dispose()
                    frm = Nothing
                End If

                If Not ReferenceEquals(cts, Nothing) Then
                    cts.Dispose()
                    cts = Nothing
                End If
            End If
        End Sub
'#End Region
    End Class
End Namespace
